### ISSUE TITLE
Slider keep in focus after it was released

### ISSUE THREAD

**Issue by ipa64** [node-red/node-red-dashboard#519]
Hi, I have noticed an annoying behavior with the slider.

### What are the steps to reproduce?
When you began to slide the cursor, the green point grow up, when you release the cursor the green point stay big, in fact the cursor stay actif, it’s stay on focus.

### What happens?
So now click anywhere on your dashboard, but outside of the slider area, for example on a button on your dashboard.
The button will be clicked and the slider cursor will get the horizontal position where you click.

### What do you expect to happen?
As soon as you release the cursor, the slider should be unfocused.

### Please tell us about your environment:

- [ ] Node-RED-Dashboard version: 2.15.5
- [ ] Node-RED version: 0.18.7
- [ ] node.js version: 8
- [ ] npm version:  ? 
- [ ] Platform/OS: IOS and MacOS 
- [ ] Browser: Safari

Thank you

**Comment by dceejay** [comment]
That seems to be the way the underlying widget library works...
for example https://material.angular.io/components/slider/examples

**Comment by ipa64** [comment]
No, in this example the behavior is correct, the cursor doesn’t move when you click elsewhere outside of the slider area ( with the focus on the slider ). On the dashboard the cursor move even when you click outside, and even when you click on another object like a button.

**Comment by dceejay** [comment]
not seeing that here I'm afraid.
Maybe need to bump Node-RED to latest version ?
![pHjnQRWWiD](https://user-images.githubusercontent.com/5375409/61540179-0b3c4a80-aa35-11e9-8404-4917e9ba181d.gif)

**Comment by ipa64** [comment]
I have just updated Nodered to 0.20.7, same behavior

Here on Safari IOS 
![SliderOnIOS](https://user-images.githubusercontent.com/10351617/61572022-80a22c80-aac2-11e9-87e8-ff2a17103fff.gif)

Here on Safari MacOS
![SliderOnMac](https://user-images.githubusercontent.com/10351617/61572171-d2978200-aac3-11e9-97f7-14aea177300c.gif)

**Comment by ipa64** [comment]
On material.angular.iothe behavior is correct, as soon the mouse button is released the cursor doesn't move anymore
![Material 2019-07-20 08_10_22](https://user-images.githubusercontent.com/10351617/61572337-21deb200-aac6-11e9-8525-b03ecaf7692b.gif)

**Comment by ipa64** [comment]
I have just tested on MacOS with the browser Google Chrome
It's working as expected !
So it's an behavior issue on Safari both on IOS and MacOS

**Comment by dceejay** [comment]
The video I posted (where it is ok) was safari in Mac. So can you share your flow.?

**Comment by ipa64** [comment]
That is very strange! I have tried on my MacBook, my Mac mini, 2 iPads, 2 iPhones.
The flow below is just a minimal flow for test.
Thank you

`[{"id":"b73f74.4270609","type":"tab","label":"Flow 1","disabled":false,"info":""},{"id":"7c176c04.f20624","type":"ui_button","z":"b73f74.4270609","name":"","group":"d9d2e0a1.937e2","order":1,"width":0,"height":0,"passthru":false,"label":"button","tooltip":"","color":"","bgcolor":"","icon":"","payload":"","payloadType":"str","topic":"","x":154,"y":109,"wires":[[]]},{"id":"2cc48ab4.2c8e46","type":"ui_slider","z":"b73f74.4270609","name":"","label":"slider","tooltip":"","group":"d9d2e0a1.937e2","order":2,"width":0,"height":0,"passthru":true,"outs":"all","topic":"","min":0,"max":10,"step":1,"x":143,"y":151,"wires":[[]]},{"id":"d9d2e0a1.937e2","type":"ui_group","z":"","name":"Default","tab":"cb993e58.e08068","disp":true,"width":"6","collapse":false},{"id":"cb993e58.e08068","type":"ui_tab","z":"","name":"Home","icon":"dashboard","order":2,"disabled":false,"hidden":false}]`

**Comment by dceejay** [comment]
hmm - sadly that is working perfectly for me.
![sVPPimSsWA](https://user-images.githubusercontent.com/5375409/61576763-cb796f80-aad5-11e9-9bcc-1ee5527637a6.gif)

**Comment by ipa64** [comment]
I don’t know, my Mac is running Mojave, Nodered is running via a docker image on my Synology NAS,  otherwise I don’t have any other idea.
Wouldn't you have access to a ios device ?
You know, for now, it’s not a big issue, I use numeric node instead slider node.

**Comment by dceejay** [comment]
No iOS Not while on vacation. 
Or you could use chrome.

**Comment by ipa64** [comment]
Yes, a workaround too
Thank you for your availability 
So, have nice vacation !

**Comment by kpm-at-hfi** [comment]
I'm seeing this on iOS Safari and Brave browsers... as @ipa64 describes, "the cursor goes to the vertical position where I push the button".

**Comment by kpm-at-hfi** [comment]
Example flow:

```[{"id":"7789fccb.ae8c54","type":"tab","label":"Flow 1","disabled":false,"info":""},{"id":"eece6bcf.9b4048","type":"ui_dropdown","z":"7789fccb.ae8c54","name":"","label":"Choose Valve","tooltip":"","place":"Select option","group":"cd676112.0da77","order":1,"width":0,"height":0,"passthru":true,"options":[{"label":"Valve 1","value":0,"type":"num"},{"label":"Valve 2","value":1,"type":"num"},{"label":"Valve 3","value":2,"type":"num"},{"label":"Valve 4","value":3,"type":"num"},{"label":"Valve 5","value":4,"type":"num"},{"label":"Valve 6","value":5,"type":"num"},{"label":"Valve 7","value":6,"type":"num"}],"payload":"","topic":"Valve Choice","x":120,"y":40,"wires":[["9f46adc.5c4905"]]},{"id":"cb73f588.64c4f8","type":"ui_button","z":"7789fccb.ae8c54","name":"","group":"cd676112.0da77","order":4,"width":0,"height":0,"passthru":false,"label":"Start","tooltip":"","color":"","bgcolor":"","icon":"","payload":"","payloadType":"date","topic":"","x":150,"y":180,"wires":[["45db6b8a.2ac79c"]]},{"id":"9f46adc.5c4905","type":"change","z":"7789fccb.ae8c54","name":"","rules":[{"t":"set","p":"valveChoice","pt":"flow","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":420,"y":40,"wires":[[]]},{"id":"f6f24241.2383d8","type":"change","z":"7789fccb.ae8c54","name":"","rules":[{"t":"set","p":"duration","pt":"flow","to":"payload","tot":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":400,"y":100,"wires":[[]]},{"id":"45db6b8a.2ac79c","type":"function","z":"7789fccb.ae8c54","name":"","func":"var chosen_valve = flow.get(\"valveChoice\")\nvar valves = [0,0,0,0,0,0,0]\nvalves[chosen_valve] = 1\n\nvar duration = flow.get(\"duration\")\n\n\n\n\n\nreturn [\n    {topic: \"Valves\", payload: valves},\n    {topic: \"Duration\", payload: duration}\n];","outputs":2,"noerr":0,"x":290,"y":180,"wires":[["7332146e.6f13c4"],["7332146e.6f13c4"]]},{"id":"7332146e.6f13c4","type":"debug","z":"7789fccb.ae8c54","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","x":570,"y":180,"wires":[]},{"id":"5ef0cfd2.cdf878","type":"ui_slider","z":"7789fccb.ae8c54","name":"Choose Duration","label":"{{value}} sec","tooltip":"","group":"cd676112.0da77","order":2,"width":0,"height":0,"passthru":true,"outs":"end","topic":"","min":"1","max":"90","step":1,"x":110,"y":100,"wires":[["f6f24241.2383d8"]]},{"id":"cd676112.0da77","type":"ui_group","z":"","name":"Default","tab":"aad4672b.88f1","disp":true,"width":"6","collapse":false},{"id":"aad4672b.88f1","type":"ui_tab","z":"","name":"Home","icon":"dashboard","disabled":false,"hidden":false}]```

Try to adjust the slider and the either "deselect" the slider to lock in the value or touch the button.

**Comment by ipa64** [comment]
Today I understand that,  in fact this problem happens with all elements, even a simple button, as soon you push it it will stay selected after it was released ( tested on iPad ). You can see it because the color stay like it was still pressed, until you push another button.

**Comment by dceejay** [comment]
The basic angular framework does not seem to recognise touchend events so this will need some digging.

**Comment by ulno** [comment]
Apologies for reading ipa64 latest comment here only just now after getting my issue marked as duplicate.

Checking https://github.com/node-red/node-red-dashboard/search?q=ng-click&type=Code reveals 9 positions where ``ng-click`` is used. Wouldn't all our touch screen problems be fixed if we replace all ``ng-click`` calls with respective calls to ``ng-touchstart``, ``ng-mousedown``, ``ng-touchend``, ``ng-mouseup`` each?

Or does the default click behavior need to be fixed upstream in angular?

I am not a good js-programmer, but fixing/rewriting the 9 positions seems pretty doable - however it will create a bit of code replication for at least two calls. And should the events be triggered on touch/mouse down or release?

I could give it a shot in a fork if there is more interest and nobody convinces me of a better approach.

**Comment by dceejay** [comment]
happy for you to have a go - but I was finding that there was no ng-touchend and if I tried to create a directive using on touchend - that still wasn't being fired by the browsers in question... but I would welcome someone else giving it a go. No need to do all 9 to start with - fixing just the slider would be a great proof of principle.

**Comment by ulno** [comment]
The quickest test one can do, is creating a button based on a template:

```html
<script>
this.scope.on = {payload: 1};
this.scope.off = {payload: 0};
</script>
<md-button
   ng-touchstart="send(on)"
   ng-mousedown="send(on)"
   ng-touchend="send(off)"
   ng-mouseup="send(off)">
       Test Button
</md-button>
```

This one works on most of my device/browser combinations, but fails on one of my older Android devices in one browser (Fully) - however that browser complained once when I started it up that the "web engine was too old" (not sure exactly about the wording - can't reproduce this error message anymore).  runs on the other more modern devices.

It would be nice if somebody could confirm that this template is indeed a workaround for some device/browser combinations.

**Comment by dceejay** [comment]
If I modify that slightly to

```
<script>
this.scope.on = {payload: 1};
this.scope.off = {payload: 0};
this.scope.on2 = {payload: 3};
this.scope.off2 = {payload: 2};
</script>
<md-button
   ng-touchstart="send(on2)"
   ng-mousedown="send(on)"
   ng-touchend="send(off2)"
   ng-mouseup="send(off)">
       Test Button
</md-button>
```

When I try that on my Iphone (ie touch) - I get nothing on touch - then 1 and 0 on release. So no touch like event fired...

**Comment by Hugobox** [comment]
So far, the only way I found was to use jquery to handle the touchstart and touchend events (as they are not natively supported by AngularJS). So far only tried for buttons but should work for sliders. see: https://discourse.nodered.org/t/momentary-button-on-dashboard/4906/5

**Comment by dceejay** [comment]
... but how to wrap that into the core widgets... ?

**Comment by ulno** [comment]
I can confirm that @Hugobox 's jquery code also works on my older Android device that does not deselect the original node-red button nor my referenced example. That's good news.

On a whim, I pulled out an old iPhone (5S updated) I still had lying around. I tried to connect with Safari, an updated Firefox, and updated Google Chrome to my node-red dashboard. None of the three solutions discussed here work on any of these browsers on iOS - I guess they are forced to all be using the same web rendering engine. So making node-red dashboard run on iOS devices might be a lost cause.

**Comment by ulno** [comment]
I made a small test fork where I adjusted the button to react on mouse-up and touch-up, now all default buttons work as expected also under touch (see below).

I updated Android System WebView on the device that didn't play well up to this point with Fully and problems are fixed now. (I also discovered a really nice new open source dashboard browser on Android which also works perfectly well with the slightly patched node-red dashboard: https://thanksmister.com/wallpanel-android/).

However I think it's not a real solution for all people present here (so I am not daring a PR at his point), but if you want to play around with the buttons and maybe soon a slider fix (working on it), install this fork on your test system: https://github.com/ulno/node-red-dashboard

**Comment by Hugobox** [comment]
AngularJS documentation mentions that while their NgTouch modules are deprecated, they recommend using HammerJS: https://hammerjs.github.io/getting-started/

**Comment by dceejay** [comment]
Last updated in 2016....

**Comment by ulno** [comment]
So, is hammerjs really the way to go forward?

**Comment by jwiesel** [comment]
Hi there,
just wanted to mention: I face the same issue as explained in the ticket description.
Node-Red: 1.0.5
Node-Red Dashboard: 2.21.0
Browser: Firefox 68.8.0 on Android
The behavior is a bit annoying with sliders to control the window blinds :see_no_evil:

Please let me know, if I can help.

**Comment by dceejay** [comment]
please try latest beta in master - hopefully now fixed (or at least better behaved)

**Comment by dceejay** [comment]
now fixed in v2.22.1
